src/basic/strv.h

```diff
@@ -122,19 +122,28 @@ static inline int strv_from_nulstr(char ***a, const char *nulstr) {
 
 bool strv_overlap(char * const *a, char * const *b) _pure_;
 
+#define _STRV_FOREACH(s, l, i)                                          \
+        for (typeof(*(l)) *s, *i = (l); (s = i) && *i; i++)
+
 #define STRV_FOREACH(s, l)                      \
-        for ((s) = (l); (s) && *(s); (s)++)
+        _STRV_FOREACH(s, l, UNIQ_T(i, UNIQ))
+
+#define _STRV_FOREACH_BACKWARDS(s, l, h, i)                             \
+        for (typeof(*(l)) *s, *h = (l),                                 \
+                     *i = h ? h + strv_length(h) - 1U : NULL;           \
+             h && (s = i) >= h;                                         \
+             i--)
```

Comment:
Not a total expert, but isn't creating a pointer that points (one element) before a malloc technically UB? (I know the code before this does this already.)

Reply:
This should be safe, but I updated this to make safer. PTAL.

Reply:
(This is still decrementing `i` and then checking, which would break for `i` pointing at the 0th string.)

Reply:
I think this is a valid point, and we should make sure to not do UB. But since this is a "preexisting condition", let's leave it for later.

Reply:
https://github.com/systemd/systemd/pull/22797